File: backend/engines/tile_hard_fix.py
"""
V91: Tile Server Hard Fix - Guaranteed Tile Load + Hard Failover
Blocking initialization, backup providers, heartbeat monitoring, zero blank maps.
"""

import httpx
import asyncio
from typing import Dict, Any, List, Optional
import random


class TileServerHeartbeat:
    """
    V91: Blocking tile server validator with hard failover.
    Ensures at least one tile server is operational before map boot.
    """
    
    def __init__(self):
        self.interval_ms = 500
        self.max_retries = 20
        self.valid_mime_types = ['image/png', 'image/jpeg', 'image/jpg', 'image/webp']
        self._client: Optional[httpx.AsyncClient] = None
        print("[V91:HEARTBEAT] Tile Server Heartbeat initialized")
    
    async def init(self):
        """Initialize HTTP client for heartbeat checks."""
        if not self._client:
            self._client = httpx.AsyncClient(
                timeout=5.0,
                follow_redirects=True,
                headers={'User-Agent': 'Pathfinder/V91'}
            )
            print("[V91:HEARTBEAT] HTTP client ready")
    
    async def test_tile_url(self, url: str) -> bool:
        """
        Test if a single tile URL is accessible and valid.
        
        Args:
            url: Complete tile URL (already formatted with z/x/y)
            
        Returns:
            True if tile is accessible and valid, False otherwise
        """
        if not self._client:
            await self.init()
        
        try:
            response = await self._client.head(url, timeout=3.0)
            if response.status_code == 200:
                return True
        except Exception as e:
            print(f"[V91:HEARTBEAT] Tile test failed for {url}: {e}")
        
        return False
    
    async def validate_with_retries(self, base_url: str) -> bool:
        """
        Validate tile server with multiple retry attempts.
        Blocks until server responds or max retries exceeded.
        
        Args:
            base_url: Tile URL template with {z}/{x}/{y} placeholders
            
        Returns:
            True if server is operational, False if all retries failed
        """
        # Create test tile URL (zoom 1, tile 1/1)
        test_url = base_url.replace('{z}', '1').replace('{x}', '1').replace('{y}', '1')
        
        print(f"[V91:HEARTBEAT] Validating: {test_url}")
        
        for attempt in range(1, self.max_retries + 1):
            if await self.test_tile_url(test_url):
                print(f"[V91:HEARTBEAT] [OK] Server operational (attempt {attempt})")
                return True
            
            if attempt < self.max_retries:
                await asyncio.sleep(self.interval_ms / 1000.0)
        
        print(f"[V91:HEARTBEAT] [FAIL] Server validation failed after {self.max_retries} attempts")
        return False
    
    async def fetch_and_validate_tile(self, url: str) -> bool:
        """
        Fetch actual tile data and validate format.
        
        Args:
            url: Complete tile URL
            
        Returns:
            True if tile data is valid (correct MIME, >32 bytes)
        """
        if not self._client:
            await self.init()
        
        try:
            response = await self._client.get(url, timeout=5.0)
            
            if response.status_code != 200:
                return False
            
            # Validate MIME type
            content_type = response.headers.get('content-type', '').lower()
            if not any(mime in content_type for mime in self.valid_mime_types):
                print(f"[V91:HEARTBEAT] Invalid MIME type: {content_type}")
                return False
            
            # Validate byte size (>32 bytes for real tile)
            if len(response.content) <= 32:
                print(f"[V91:HEARTBEAT] Tile too small: {len(response.content)} bytes")
                return False
            
            print(f"[V91:HEARTBEAT] [OK] Tile valid: {len(response.content)} bytes, {content_type}")
            return True
            
        except Exception as e:
            print(f"[V91:HEARTBEAT] Tile fetch error: {e}")
            return False
    
    async def close(self):
        """Close HTTP client."""
        if self._client:
            await self._client.aclose()
            print("[V91:HEARTBEAT] HTTP client closed")


class TileFallbackProvider:
    """
    V91: Backup tile provider system with auto-failover.
    Maintains list of fallback servers for guaranteed tile availability.
    """
    
    def __init__(self):
        self.providers = [
            {
                'name': 'OpenStreetMap',
                'url': 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
                'priority': 1
            },
            {
                'name': 'Carto Light',
                'url': 'https://a.basemaps.cartocdn.com/light_all/{z}/{x}/{y}.png',
                'priority': 2
            },
            {
                'name': 'HOT OSM',
                'url': 'https://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png',
                'priority': 3
            },
            {
                'name': 'Carto Dark',
                'url': 'https://basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png',
                'priority': 4
            }
        ]
        print(f"[V91:FALLBACK] Initialized with {len(self.providers)} backup providers")
    
    def select_random(self) -> Dict[str, Any]:
        """Select random backup provider."""
        provider = random.choice(self.providers)
        print(f"[V91:FALLBACK] Selected: {provider['name']}")
        return provider
    
    def select_by_priority(self) -> Dict[str, Any]:
        """Select highest priority backup provider."""
        provider = sorted(self.providers, key=lambda x: x['priority'])[0]
        print(f"[V91:FALLBACK] Selected priority: {provider['name']}")
        return provider
    
    def get_all_urls(self) -> List[str]:
        """Get all backup provider URLs."""
        return [p['url'] for p in self.providers]


class TileFormatEnforcer:
    """
    V91: Tile format validator to prevent broken/empty tiles.
    Validates MIME types and byte sizes before rendering.
    """
    
    def __init__(self):
        self.valid_mime_types = ['image/png', 'image/jpeg', 'image/jpg', 'image/webp']
        self.min_bytes = 32
        print("[V91:ENFORCER] Tile Format Enforcer initialized")
    
    def verify_mime(self, content_type: str) -> bool:
        """
        Verify content-type header is valid image format.
        
        Args:
            content_type: HTTP Content-Type header value
            
        Returns:
            True if valid image MIME type
        """
        content_type_lower = content_type.lower()
        is_valid = any(mime in content_type_lower for mime in self.valid_mime_types)
        
        if not is_valid:
            print(f"[V91:ENFORCER] [FAIL] Invalid MIME: {content_type}")
        
        return is_valid
    
    def verify_bytes(self, data: bytes) -> bool:
        """
        Verify tile data has sufficient bytes (not empty/broken).
        
        Args:
            data: Raw tile bytes
            
        Returns:
            True if data size exceeds minimum threshold
        """
        is_valid = len(data) > self.min_bytes
        
        if not is_valid:
            print(f"[V91:ENFORCER] [FAIL] Tile too small: {len(data)} bytes")
        
        return is_valid
    
    def validate_tile(self, content_type: str, data: bytes) -> bool:
        """
        Full tile validation (MIME + bytes).
        
        Args:
            content_type: HTTP Content-Type header
            data: Raw tile bytes
            
        Returns:
            True if tile passes all validation checks
        """
        return self.verify_mime(content_type) and self.verify_bytes(data)


# Global instances
_tile_heartbeat: Optional[TileServerHeartbeat] = None
_tile_fallback: Optional[TileFallbackProvider] = None
_tile_enforcer: Optional[TileFormatEnforcer] = None


def get_tile_heartbeat() -> TileServerHeartbeat:
    """Get global TileServerHeartbeat instance."""
    global _tile_heartbeat
    if _tile_heartbeat is None:
        _tile_heartbeat = TileServerHeartbeat()
    return _tile_heartbeat


def get_tile_fallback() -> TileFallbackProvider:
    """Get global TileFallbackProvider instance."""
    global _tile_fallback
    if _tile_fallback is None:
        _tile_fallback = TileFallbackProvider()
    return _tile_fallback


def get_tile_enforcer() -> TileFormatEnforcer:
    """Get global TileFormatEnforcer instance."""
    global _tile_enforcer
    if _tile_enforcer is None:
        _tile_enforcer = TileFormatEnforcer()
    return _tile_enforcer

File: frontend/src/services/tileHardFix.ts
/**
 * PATHFINDER V91 — TILE SERVER HARD FIX (Frontend)
 * Guaranteed tile load, hard retry logic, blocking initialization
 */

interface TileLoadResult {
  success: boolean;
  data: ArrayBuffer | null;
  url: string;
  attempts: number;
  error?: string;
}

interface HeartbeatResult {
  primaryValid: boolean;
  fallbackUrl?: string;
  message: string;
}

class TileHardLoader {
  private maxAttempts = 5;
  private retryDelay = 200; // ms
  private validMimeTypes = ['image/png', 'image/jpeg', 'image/jpg', 'image/webp'];
  private minBytes = 32;

  /**
   * Hard tile loader - retries until success or max attempts
   * @param url Complete tile URL
   * @returns TileLoadResult with data or null
   */
  async loadTileHard(url: string): Promise<TileLoadResult> {
    let attempt = 0;

    while (attempt < this.maxAttempts) {
      attempt++;
      
      try {
        const response = await fetch(url, {
          method: 'GET',
          cache: 'default',
          headers: {
            'User-Agent': 'Pathfinder/V91'
          }
        });

        if (response.status === 200) {
          const contentType = response.headers.get('content-type') || '';
          const buffer = await response.arrayBuffer();

          // Validate MIME type
          if (!this.validMimeTypes.some(mime => contentType.includes(mime))) {
            console.warn(`[V91:HARD_LOADER] Invalid MIME: ${contentType}`);
            continue;
          }

          // Validate byte size
          if (buffer.byteLength <= this.minBytes) {
            console.warn(`[V91:HARD_LOADER] Tile too small: ${buffer.byteLength} bytes`);
            continue;
          }

          console.log(`[V91:HARD_LOADER] [OK] Tile loaded (attempt ${attempt}): ${buffer.byteLength} bytes`);
          return {
            success: true,
            data: buffer,
            url,
            attempts: attempt
          };
        }
      } catch (error) {
        console.warn(`[V91:HARD_LOADER] Attempt ${attempt} failed:`, error);
      }

      // Wait before retry (except last attempt)
      if (attempt < this.maxAttempts) {
        await this.wait(this.retryDelay);
      }
    }

    console.error(`[V91:HARD_LOADER] [FAIL] All ${this.maxAttempts} attempts failed for ${url}`);
    return {
      success: false,
      data: null,
      url,
      attempts: this.maxAttempts,
      error: 'Max retries exceeded'
    };
  }

  /**
   * Test if a tile URL is accessible
   * @param url Complete tile URL
   * @returns true if tile is accessible
   */
  async testTileUrl(url: string): Promise<boolean> {
    try {
      const response = await fetch(url, {
        method: 'HEAD',
        cache: 'no-cache'
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  private wait(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

class TileServerHeartbeat {
  private backendUrl = 'http://localhost:8000';
  private heartbeatInterval = 3000; // ms
  private intervalId: number | null = null;
  private onServerSwitch?: (message: string) => void;

  constructor(onServerSwitch?: (message: string) => void) {
    this.onServerSwitch = onServerSwitch;
  }

  /**
   * Check backend tile heartbeat status
   * @returns HeartbeatResult with server status
   */
  async checkHeartbeat(): Promise<HeartbeatResult> {
    try {
      const response = await fetch(`${this.backendUrl}/api/v1/heartbeat/tiles`);
      const data = await response.json();

      if (data.primary_server?.operational) {
        return {
          primaryValid: true,
          message: 'Primary tile server operational'
        };
      } else {
        const fallbackUrl = data.fallback_providers?.[0];
        return {
          primaryValid: false,
          fallbackUrl,
          message: 'Primary server offline - fallback required'
        };
      }
    } catch (error) {
      console.error('[V91:HEARTBEAT] Backend heartbeat check failed:', error);
      return {
        primaryValid: false,
        message: 'Heartbeat check failed - network error'
      };
    }
  }

  /**
   * Start live heartbeat monitoring
   * Checks tile server every 3 seconds and switches on failure
   */
  startMonitoring(): void {
    if (this.intervalId !== null) {
      console.warn('[V91:HEARTBEAT] Monitoring already active');
      return;
    }

    console.log('[V91:HEARTBEAT] Starting live tile heartbeat monitor');

    this.intervalId = window.setInterval(async () => {
      const result = await this.checkHeartbeat();

      if (!result.primaryValid) {
        console.warn('[V91:HEARTBEAT] Server heartbeat failed - switching provider');
        this.onServerSwitch?.(result.message);
      }
    }, this.heartbeatInterval);
  }

  /**
   * Stop heartbeat monitoring
   */
  stopMonitoring(): void {
    if (this.intervalId !== null) {
      clearInterval(this.intervalId);
      this.intervalId = null;
      console.log('[V91:HEARTBEAT] Heartbeat monitoring stopped');
    }
  }
}

class BlockingMapInitializer {
  private hardLoader = new TileHardLoader();
  private heartbeat: TileServerHeartbeat;
  private backendUrl = 'http://localhost:8000';

  constructor(onServerSwitch?: (message: string) => void) {
    this.heartbeat = new TileServerHeartbeat(onServerSwitch);
  }

  /**
   * Initialize map with blocking validation
   * Map will NOT render until tiles are verified
   * @returns true if initialization successful, false otherwise
   */
  async initializeBlocking(): Promise<boolean> {
    console.log('[V91:INIT] Starting blocking map initialization...');

    // Step 1: Check backend heartbeat
    const heartbeatResult = await this.heartbeat.checkHeartbeat();
    console.log(`[V91:INIT] Heartbeat: ${heartbeatResult.message}`);

    // Step 2: Test actual tile fetch
    const testTileUrl = 'https://basemaps.cartocdn.com/dark_all/2/2/2.png';
    console.log(`[V91:INIT] Testing tile fetch: ${testTileUrl}`);

    const tileResult = await this.hardLoader.loadTileHard(testTileUrl);

    if (tileResult.success) {
      console.log('[V91:INIT] [OK] Tile test PASSED - map ready to initialize');
      
      // Step 3: Start live heartbeat monitoring
      this.heartbeat.startMonitoring();
      
      return true;
    } else {
      console.error('[V91:INIT] [FAIL] Tile test FAILED - map initialization blocked');
      
      // Try fallback if available
      if (heartbeatResult.fallbackUrl) {
        console.log('[V91:INIT] Attempting fallback provider...');
        const fallbackUrl = heartbeatResult.fallbackUrl.replace('{z}', '2').replace('{x}', '2').replace('{y}', '2');
        const fallbackResult = await this.hardLoader.loadTileHard(fallbackUrl);
        
        if (fallbackResult.success) {
          console.log('[V91:INIT] [OK] Fallback tile test PASSED');
          this.heartbeat.startMonitoring();
          return true;
        }
      }
      
      console.error('[V91:INIT] [FAIL] All tile tests failed - map cannot initialize');
      return false;
    }
  }

  /**
   * Stop all monitoring
   */
  cleanup(): void {
    this.heartbeat.stopMonitoring();
  }

  /**
   * Get hard loader instance for manual tile loading
   */
  getHardLoader(): TileHardLoader {
    return this.hardLoader;
  }
}

// Singleton instances
let _tileHardLoader: TileHardLoader | null = null;
let _blockingInitializer: BlockingMapInitializer | null = null;

export function getTileHardLoader(): TileHardLoader {
  if (!_tileHardLoader) {
    _tileHardLoader = new TileHardLoader();
  }
  return _tileHardLoader;
}

export function getBlockingMapInitializer(
  onServerSwitch?: (message: string) => void
): BlockingMapInitializer {
  if (!_blockingInitializer) {
    _blockingInitializer = new BlockingMapInitializer(onServerSwitch);
  }
  return _blockingInitializer;
}

export { TileHardLoader, TileServerHeartbeat, BlockingMapInitializer };
export type { TileLoadResult, HeartbeatResult };
